### DISCUSSION TITLE
9 to 10 migration with Next, .interop() and t.mergeRouters?

### DISCUSSION THREAD

**Discussion by ftzi** [trpc/trpc#3449] [Q&A]
I am doing the migration of trpc@9 to 10 in my project that uses Next.js.

I added `.interop` to our router as the PR got quite big with all endpoints migrated. So it ended up like this:

```tsx
const legacyRouter = createMyRouter()
  .transformer(mySuperjsonTransformer)
  .merge(myRouter1)
  .merge(myRouter2)
  .interop()
  
const appRouter = legacyRouter
```

And as we use Next, we kept consuming it with `trpc.useQuery('a.b')` syntax as [it's said the new syntax won't work with interop](https://trpc.io/docs/nextjs#4-create-trpc-hooks).

It worked. But, as soon as I tried adding the mainRouter with mergeRouters, everything stopped working:

```tsx
const mainRouter = t.router({
  greeting: t.procedure.query(() => 'hello from tRPC v10!'),
})

// Merge v9 router with v10 router
export const appRouter = t.mergeRouters(legacyRouter, mainRouter)
```

Also, 'greeting' isn't accessible via `trpc.useQuery(['greeting'])` (it doesn't show up there) and we can't use trpc.greeting as again, it's said Next won't work with interop and we still use it.

Am I missing something here? Isn't the interop partial migration possible with Next.js?

**Comment by KATT** [comment]
IIRC it's `trpc.proxy.greeting.*`. Maybe you have to create the proxy client separately. In the React-one I think we add `trpc.proxy.*` automatically but I don't remember exactly. You can also dig around [the tests](https://github.com/trpc/trpc/tree/main/packages/tests/server/interop) to see if they bring you any clarity.

Anyway, I highly recommend you to use @sachinraja's [great codemod](https://github.com/sachinraja/trpc-v10-migrate-codemod) and do it all in one go, it'll be worth it.

**Reply by ftzi** [reply]
I used the codemod (and the required manual changes) in the first PR I did for the migration but it was quite hard to review it as there were many changes, so we are trying to do partial migrations to better analyze it and avoid small mistakes. If we find out it's not possible we will do that quite big PR as we won't have another way to go (and we want to upgrade!)

I just found out the `error - Error: You seem to have several transformers` that was showing up in my Next console and removed the transformer in my `initTRPC.create`. It seems that it didn't like that I used it both in my legacyRouter and in this mainRouter, so now it's working!

**Reply by lukevella** [reply]
I've just encountered the same issue and can't seem to find a way around this. 

If you only add the transformer to the v10 router you get ts errors saying `You must set a transformer on the backend router`.

If you add the transformer to both the v9 router and the v10 router silences all the ts errors but you get the error `You seem to have several transformers`.

If you only add the transformer to the v9 router then the transformer is missing on the new routes which results in ts errors.

Here's my branch if it helps: https://github.com/lukevella/rallly/pull/489

**Reply by ftzi** [reply]
@lukevella I was using a fake transformer in v10 (I already completed the full migration).

I was doing something like { tranformer: {} as typeof v9Tranformer }, while the v9 was using the real transformer. The v10 will use v9 transformer, but we need to pass this type to v10. I don't recall exactly how it was, but just find a way to pass the transformer type to v10 and keep the real one in v9.

**Reply by KATT** [reply]
Is it really that hard to review? Do you know about the hide whitespace button when reviewing? 😅   ![CleanShot 2023-02-07 at 19 53 15](https://user-images.githubusercontent.com/459267/217339092-334da5d5-10ac-46fb-9ebc-e617fcae4bee.png)

Had a stab on @lukevella's PR https://github.com/lukevella/rallly/pull/491

**Reply by ftzi** [reply]
The problem with the migration is that in my project it was required to change a lot of stuff, so a single PR didn't make it, it was too complex and had many breaking things. Indeed doing a migration over many PRs was the best, as many underlying issues were found, also due to the fact that we had some quite complex utils around trpc, like wrappers around queries and mutations to automatically insert some params.

Now we have even a proxy and some good TS magic around tRPC to add some stuff methods to the procedures, like trpc.a.b.useQueryWithX(), that only shows up if the procedure has the right input prop type.

The codemod almost did it, but if it was used, it also required those many changes to be made at the same point, else the CI/CD would be broken.

Anyway, no longer a problem for me as it's over already
